File: algorithmic_toolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/FractionalKnapsack.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

class Item {
  public int weight;
  public int value;
  public double unitValue;
}

public class FractionalKnapsack {
  private static double getOptimalValue(int capacity, int[] values, int[] weights) {
    int weight = 0;
    double value = 0.0;

    List<Item> items = new ArrayList<>();
    for (int i = 0; i < weights.length; i++) {
      Item item = new Item();
      item.weight = weights[i];
      item.value = values[i];
      item.unitValue = (double) values[i] / weights[i];
      items.add(item);
    }

    Collections.sort(items, (a, b) -> Double.compare(a.unitValue, b.unitValue));
    Collections.reverse(items);

    for (Item item : items) {
      if (weight >= capacity)
        break;

      int units = Math.min(item.weight, capacity - weight);
      weight += units;
      value += units * item.unitValue;
    }

    return value;
  }

  public static void main(String args[]) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();
    int capacity = scanner.nextInt();
    int[] values = new int[n];
    int[] weights = new int[n];
    for (int i = 0; i < n; i++) {
      values[i] = scanner.nextInt();
      weights[i] = scanner.nextInt();
    }
    scanner.close();

    System.out.println(getOptimalValue(capacity, values, weights));
  }
}

File: algorithmic_toolbox/week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.py
import sys
from typing import Sequence


def get_optimal_value(*, capacity: int, weights: Sequence[int], values: Sequence[int]) -> float:
    weight = 0
    value = 0.0

    for item in sorted(
        [{"weight": weight, "value": value, "unit-value": value / weight} for weight, value in zip(weights, values)],
        key=lambda item: item["unit-value"],
        reverse=True,
    ):
        if weight >= capacity:
            break

        weight += (units := int(min(item["weight"], capacity - weight)))
        value += units * item["unit-value"]

    return value


if __name__ == "__main__":
    data = list(map(int, sys.stdin.read().split()))
    n, capacity = data[0:2]
    values = data[2 : (2 * n + 2) : 2]
    weights = data[3 : (2 * n + 2) : 2]
    opt_value = get_optimal_value(capacity=capacity, weights=weights, values=values)
    print("{:.10f}".format(opt_value))
